pkg/provider/bao/bao.go

```diff
@@ -0,0 +1,174 @@
+	// it is not a BAO_* variable.
+	// Additionally, in a login scenario, we include BAO_* variables in the secrets list.
+	if !ok || (s.login && envType.login) {
+		// Path here is actually the secret's key,
+		// An example of this can be found at the LoadSecrets() function below
+		secret := provider.Secret{
+			Path:  key,
+			Value: value,
+		}
+
+		s.secrets = append(s.secrets, secret)
+	}
+}
+
+func NewProvider(config *Config) (*Provider, error) {
+	clientOptions := []bao.ClientOption{bao.ClientLogger(clientLogger{slog.Default()})}
+	if config.TokenFile != "" {
+		clientOptions = append(clientOptions, bao.ClientToken(config.Token))
+	} else {
+		// use role/path based authentication
+		clientOptions = append(clientOptions,
+			bao.ClientRole(config.Role),
+			bao.ClientAuthPath(config.AuthPath),
+			bao.ClientAuthMethod(config.AuthMethod),
+		)
+	}
+
+	client, err := bao.NewClientWithOptions(clientOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create bao client: %w", err)
+	}
+
+	// Accessing the application configuration is necessary
+	// to determine whether daemon mode is enabled
+	appConfig, err := common.LoadConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load application config: %w", err)
+	}
+
+	injectorConfig := injector.Config{
+		TransitKeyID:         config.TransitKeyID,
+		TransitPath:          config.TransitPath,
+		TransitBatchSize:     config.TransitBatchSize,
+		IgnoreMissingSecrets: config.IgnoreMissingSecrets,
+		DaemonMode:           appConfig.Daemon,
+	}
+
+	var secretRenewer injector.SecretRenewer
+
+	if appConfig.Daemon {
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs)
+
+		secretRenewer = daemonSecretRenewer{client: client, sigs: sigs}
+		slog.Info("Daemon mode enabled. Will renew secrets in the background.")
+	}
+
+	return &Provider{
+		isLogin:        config.IsLogin,
+		client:         client,
+		injectorConfig: injectorConfig,
+		secretRenewer:  secretRenewer,
+		fromPath:       config.FromPath,
+		revokeToken:    config.RevokeToken,
+	}, nil
+}
+
+// LoadSecret's path formatting: <key>=<path>
+// This formatting is necessary because the injector expects a map of key=value pairs.
+// It also returns a map of key:value pairs, where the key is the environment variable name
+// and the value is the secret value
+// E.g. paths: MYSQL_PASSWORD=secret/data/mysql/password
+// returns: []provider.Secret{provider.Secret{Path: "MYSQL_PASSWORD", Value: "password"}}
+func (p *Provider) LoadSecrets(_ context.Context, paths []string) ([]provider.Secret, error) {
+	sanitized := sanitized{login: p.isLogin}
+	baoEnviron := parsePathsToMap(paths)
+
+	secretInjector := injector.NewSecretInjector(p.injectorConfig, p.client, p.secretRenewer, slog.Default())
+	inject := func(key, value string) {
+		sanitized.append(key, value)
+	}
+
+	err := secretInjector.InjectSecretsFromBao(baoEnviron, inject)
+	if err != nil {
+		return nil, fmt.Errorf("failed to inject secrets from bao: %w", err)
+	}
```

Comment:
should this injection be executed when we specify from path?

Reply:
Yes, but It has been fixed in the latest commit, now its like this:
```go
	if p.fromPath != "" {
		err = secretInjector.InjectSecretsFromBaoPath(p.fromPath, inject)
		if err != nil {
			return nil, fmt.Errorf("failed to inject secrets from bao path: %w", err)
		}
	}
```

`InjectSecrtesFromBao` uses the whole environ.
`InjectSecretsFromBaoPath` uses the `BAO_FROM_PATH` env-var.